### PR TITLE
Feature/include fcids in aggregate report

### PR DESCRIPTION
**Pull request by b97pla**
- Parses the flowcell id, if available, from the setup XML files and includes in the aggregate csv report
- Enables passing additional sample information on the command line.

### CODE REVIEW

**Comment by senthil10** [comment]
@ewels : looks fine to me, but I dont know much about `piper` xml files. So if fine by you I can merge it (than I can make by *BIG PR* :smiling_imp:  ).

**Comment by ewels** [comment]
Sorry @senthil10 - got halfway through looking through it earlier and got distracted. Will continue now.

File: ngi_reports/common/__init__.py [hunk 1]
```diff
@@ -29,6 +30,26 @@ def __init__(self, config, LOG, working_dir, **kwargs):
     def parse_piper_xml(self):
         """ Parses the XML setup files that Piper uses
         """
+        
+        def _nextitem(xml):
```
Comment (ewels) [inline ngi_reports/common/__init__.py]: Much nicer :+1: My implementation made me squirm a little, I figured that there should be a better way to do it..

File: ngi_reports/common/__init__.py [hunk 2]
```diff
@@ -58,18 +79,16 @@ def parse_piper_xml(self):
...
+                    for sample in _nextitem(run['inputs']['sample']):
+                        try:
+                            fcid = [_getfcid(pfu['unitinfo']) \
+                                for library in _nextitem(sample['library']) \
+                                for pfu in _nextitem(library['platformunit'])]
+                        except (IndexError, KeyError) as e:
+                            self.LOG.warning('Could not parse platform unit info in sample XML file: {}'.format(e.message))
+                            fcid = ['N/A']
+                        sid = sample['samplename']
+                        samples[sid] = {'id': sid, 'run_id': ';'.join(set(fcid))}
```
Comment (ewels) [inline ngi_reports/common/__init__.py]: What's the reason for logging `'id': sid` as well as using `sid` in the dict key? Just convienience when reading the dict? eg. Why not use `samples[sid] = fcid`?

File: scripts/ngi_reports [hunk 3]
```diff
@@ -151,6 +152,7 @@ if __name__ == "__main__":
     parser.add_argument('report_type', choices=allowed_report_types, metavar='<report type>', help="Type of report to generate. Choose from: {}".format(', '.join(allowed_report_types)))
     parser.add_argument('--project', default=None, action="store", help="Project name to generate 'project_summary' report")
     parser.add_argument('--samples', default=None, action="store", nargs="*", help="Limit the samples to include in reports")
+    parser.add_argument('--samples_extra', default=None, action="store", type=json.loads, help="Pass in extra information about samples as a json string, having each sample as a key. Example: --samples_extra '{\"TS001-1\": {\"delivered\": \"20150701\"}}'")
```
Comment (ewels) [inline scripts/ngi_reports]: Nice!

**Comment by ewels** [comment]
Yeah, looks awesome as ever.

File: ngi_reports/common/__init__.py (hunk 2 above)
Comment (b97pla) [inline ngi_reports/common/__init__.py]: I think the strongest reason is to have the id field conveniently end up as a column in the aggregated csv report but it's also used in at least one other place.

**Pull request merged**